### ISSUE TITLE
[google_maps_flutter_web] Are events handled correctly in Geometry controllers?

### ISSUE THREAD

**Issue by ditman** [flutter/flutter#64084]
Should the `consumeTapEvents` property on Geometry controllers prevent events from being added to the shared Stream? Do we need to prevent propagation of these events differently?

**Comment by Sammius** [comment]
Hello, @ditman!
Your question was asked a long time ago, but I just now ran into this problem. For markers, this was not a big problem, but now I started using polygons and it now looks critical to me.
It would be nice if the behavior on the web was the same as on the apps. In applications, when I set the `consumeTapEvents: false` it becomes "transparent" for taps.

**Comment by ditman** [comment]
> In applications, when I set the consumeTapEvents: false it becomes "transparent" for taps.

@Sammius this is very good feedback, thanks for chiming in. What do you mean by "transparent" does it not even fire the tap event?

**Comment by Sammius** [comment]
> What do you mean by "transparent" does it not even fire the tap event?

In this case, the onTap event of the google map itself is triggered in the application.

**Comment by ditman** [comment]
> ignores the event on the polygon and fires the `onTap` event of the Google map itself.

Very interesting! This seems very doable in web, but this needs more investigation on how to determine the best way of doing this in the web (Map onTap comes with a LatLng, which I'm not 100% sure how to guess in the web)

**Comment by Sammius** [comment]
> Very interesting!

In my opinion, the behavior of this parameter in the application is a bit strange:). But I'd like everything to be on the web exactly the same as in the application:

For markers:
>True if the marker icon consumes tap events. If not, the map will perform default tap handling by centering the map on the marker and displaying its info window.

For polylines and polygons:
>True if the [Polyline/Polygon] consumes tap events.
>If this is false, [onTap] callback will not be triggered.

**Comment by ditman** [comment]
> Currently, there is NO WAY to get onTap() nor onLongPress() events in the GoogleMapsFlutter[WEB] when using polygons.

Polygons should have an `onTap` constructor parameter that should allow this, however (just by reading at the code), I think it might not be being used at all.

This is where most of the polygon options are converted to gmaps':

* https://github.com/flutter/plugins/blob/a87aad1458a6d461c7a70d64fddafe3c925fdca8/packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart#L320-L329

This is where the code above is used:

* https://github.com/flutter/plugins/blob/a87aad1458a6d461c7a70d64fddafe3c925fdca8/packages/google_maps_flutter/google_maps_flutter_web/lib/src/polygons.dart#L39-L49

It seems to me that the user-supplied onTap parameter is never hooked up to the underlying polygon, but only our own synthetic event:

https://github.com/flutter/plugins/blob/a87aad1458a6d461c7a70d64fddafe3c925fdca8/packages/google_maps_flutter/google_maps_flutter_web/lib/src/polygon.dart#L17-L19

Some changes that are needed here:

* If the user sets the polygon as `_consumeTapEvents = false`, the polygon should be `clickable = false`, and not absorb any click events (I guess letting them bubble up as a "map" click). See [PolygonOptions in gmaps' docs](https://developers.google.com/maps/documentation/javascript/reference/polygon#PolygonOptions.clickable). This should assert that onTap is null too, to let the user know that they can't specify both _consumeTapEvents = false AND a callback, it won't work.
* If `_consumeTapEvents = true` and the user does NOT provide a callback, we should continue funneling the events to the event system. We should attempt to gather more information from the JS event. (This is the current behavior).
* If the user does provide a callback, we should continue funneling the events to the event system AND call the callback, so the onTap works as expected.

In order to synthesize the onLongPress, we could hook the [`contextmenu` event](https://developers.google.com/maps/documentation/javascript/reference/polygon#Polygon.contextmenu) so the right click approximates the longpress (as the Map does).

**Comment by stenlee** [comment]
@ditman - thank you for your feedback. Yes - the changes you described are needed to make this work.

- the Polygons (and Circles) represent much larger areas than Markers or Polylines. Because of this, information about the LatLng of user interaction is required in many use cases. But LatLng is currently available only in the top-level callbacks of the GoogleMap widget.

- In google maps for Flutter (iOS+Android), the "click transparency"  is working as expected and onTap(LatLng) + onLongPress(LatLng) is fired as expected when consumeTapEvents is set to false for Polygon or Circle.
- additional improvement I would like to see is adding the LatLng information to the onTap() callback for Polygons and Circles.


I played with your source code and I was able to achieve some results with the following changes

- https://github.com/stenlee/plugins/blob/main/packages/google_maps_flutter/google_maps_flutter_web/lib/src/polygons.dart#L43-L57
- https://github.com/stenlee/plugins/blob/main/packages/google_maps_flutter/google_maps_flutter_web/lib/src/polygon.dart#L10-L30

**Comment by ditman** [comment]
> additional improvement I would like to see is adding the LatLng information to the onTap() callback for Polygons and Circles.

@stenlee I agree. IMO on the web this is totally doable, since the [PolyMouseEvent](https://developers.google.com/maps/documentation/javascript/reference/3.48/polygon#PolyMouseEvent) extends [MapMouseEvent](https://developers.google.com/maps/documentation/javascript/reference/3.48/map#MapMouseEvent) (which contains the `latLng` of the click).

I'm not sure about other platforms. Both [iOS](https://stackoverflow.com/a/39990058), and [Android](https://developers.google.com/maps/documentation/android-sdk/shapes?hl=en#polygon_events) seem _hacky_.

_(This predates me joining the Flutter team, but I'd say this bit of the plugin API was created based off of the Android SDK way back then, and that's why it only passes a Polygon back; we've seen similar decisions to this in the past!)_

**Comment by raphael-bmec-co** [comment]
+1 for this issue. With `Polygon`s always catching clicks - whether `onTap` and/or `consumeTapEvents` - we effectively cannot catch any other map taps because there is a possibility of a Polygon 😵‍💫 .

**Comment by raphael-bmec-co** [comment]
Hi team. There has been a few months of inactivity here. I don't believe this should have a `new feature` tag. This is exposed API that is not working on web.

**Comment by ditman** [comment]
> Hi team. There has been a few months of inactivity here. I don't believe this should have a `new feature` tag. This is exposed API that is not working on web.

@raphael-bmec-co fair, I've re-tagged this as "parity" rather than "new feature".

**Comment by dimonkomaron** [comment]
Hi @ditman ! Any updates or roadmap for this issue or quick solution that can be added to fork?

**Comment by dimonkomaron** [comment]
Hi! Is there progress, roadmap or workaround possible to overcome this issue?
